Rewrite Resource Controller docs (#5320)
Qiskit/documentation: docs/guides/access-instances-platform-apis.mdx

---
title: Use IBM Cloud Resource Controller API for instance management
description: Use the IBM Cloud Resource Controller REST API to get, list, update, and create Qiskit Runtime instances.

---

{/* cspell:ignore Abluemix */}

# Use IBM Cloud Resource Controller API for instance management

You can use the IBM Cloud&reg; Resource Controller REST API to programmatically get, create, and update [instances](/docs/guides/instances). 

All Resource Controller endpoints require that you authenticate by passing a header called `Authorization` with the bearer token. Refer to the [REST API setup guide](/docs/guides/cloud-setup-rest-api).

## Get an instance

Use the [`GET /v2/resource_instances/{crn}`](https://cloud.ibm.com/apidocs/resource-controller/resource-controller#get-resource-instance) endpoint to get information about a particular instance. The [CRN](/docs/guides/instances) must be URL-encoded in the path.

> In addition to the standard Resource Controller fields, the response includes quantum-specific fields in `extensions`. Use `extensions` rather than `parameters` because `parameters` might include a subset of the full fields.

Reviewer: I feel like it would be good to mention `parameters` here in the response, to explain what they are, but then be done with them and only refer to `extensions` later in this doc. Also, the issue is not only that `parameters` are a subset, but the actual data that gets written (by the Service Broker) can be different than what it was passed, and extensions should have the current configuration that was stored by the service.

Maybe something like this?

The response includes quantum-specific fields in both `parameters` and `extensions`. Read configuration from `extensions` rather than `parameters`: `parameters` echoes your most recent request, while `extensions` reflect the configuration that was stored by the service.

The `extensions` object includes these fields:

- `instance_limit_seconds` — Integer, or `null`. The usage time limit for the instance. See [Set instance allocation limits](/docs/guides/allocation-limits).
- `usage_allocation_seconds` — Integer, or `null`. The time allocated to this instance, used by the [fair-share scheduler](/docs/guides/fair-share-scheduler) to determine queue priority. See [Set instance allocation limits](/docs/guides/allocation-limits).
- `backends` — Array of strings. The allowlist of backend names available to this instance. `["ANY"]` means all backends on the plan are available (the default). `[]` means no backends are available.

<Admonition type="caution" title="Some extensions fields may be outdated">

>   The following fields in the `extensions` object may be outdated. This can happen when IBM Quantum Support changes your instance's configuration. The current workaround is to instead use the [Qiskit Runtime REST API](/docs/api/qiskit-runtime-rest) to get this metadata.

Reviewer: I wonder if we can update this wording: the IBM Quantum Support team doesn't directly change an instance's configuration — it might seem like they have permission to do that, which they don't. Can we say something like (this may be too wordy, but its here for the idea): This can happen when an account update made by IBM Quantum Support results in a change to your instance, for example when a backend is removed from an account, it is automatically removed from an instance of the account.

>   The following fields in the `extensions` object may be outdated. This can happen when IBM Quantum Support changes your instance's configuration. The current workaround is to instead use the [Qiskit Runtime REST API](/docs/api/qiskit-runtime-rest) to get this metadata.

Reviewer: Minor nit:
> The following fields
We plan to add two more fields in the near future (functions, and custom_functions), but for now, should this be singular?

Author: Yeah, it probably should be singular for now. I indeed was trying to future-proof it for your upcoming docs, but that's confusing to users

>   - `backends`. Instead, use [`GET /v1/instance`](/docs/api/qiskit-runtime-rest/tags/instances#tags__instances__operations__get_instance).

Reviewer: The endpoint to use for backends is `GET /v1/backends`

Author: Nack. The `/v1/backends` endpoint returns account-level info. The instances endpoint returns the backends for the specific instance

```
{
  "backends": [
    "ibm_backend"
  ],
  "instance_limit": 60.7,
  "instance_limit_seconds": "61",
  "plan_id": "flex-plan-id-1",
  "usage_allocation_seconds": "200"
}
```

See https://quantum.cloud.ibm.com/docs/en/api/qiskit-runtime-rest/tags/instances#tags__instances__operations__get_instance

Reviewer: `GET /v1/backends` requires a `Service-CRN` header. It returns the backends that are configured for the instance that is passed in the header.

The endpoints that returns backends for an account is `GET /accounts/:id`.

Author: Oh, hm. Thanks for pointing that out. That being said, is there any issue with using `GET /v1/instance`? I like it because it includes the other metadata that the Resource Controller would return: limit and allocation

>   Instead, the current workaround is to use the [Qiskit Runtime REST API](/docs/api/qiskit-runtime-rest) with the [`GET /v1/backends`](/docs/api/qiskit-runtime-rest/backends#tags__backends__operations__list_backends) endpoint. (Make sure that you set the `Service-CRN` header to your instance's CRN.)

Reviewer 2: ```suggestion
  Instead, the current workaround is to use the [Qiskit Runtime REST API](/docs/api/qiskit-runtime-rest) with the [`GET /v1/backends`](/docs/api/qiskit-runtime-rest/tags/backends) endpoint. (Make sure that you set the `Service-CRN` header to your instance's CRN.)
```
Something funky is going on with our anchors on that page and needs further investigation. This URL should work in the meantime, I think

Author: The issue isn't the anchor, but the bad URL path

    import requests

    crn = "<YOUR_INSTANCE_CRN>"
    # We use urllib.parse.quote to URL-encode the CRN.
    url = (
        "https://resource-controller.cloud.ibm.com/v2/resource_instances/"
        + urllib.parse.quote(crn, safe="")
    )

    resp = requests.get(
        url,
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    resp.raise_for_status()
    print(resp.json())
    ```
  </TabItem>
</Tabs>

## Get a list of all instances

Use the [`GET /v2/resource_instances`](https://cloud.ibm.com/apidocs/resource-controller/resource-controller#list-resource-instances) endpoint to get a list of all your instances. Set the `resource_id` query parameter to `b6049020-80f4-11eb-a0f7-e35ec9b4054f` to filter to IBM Quantum&reg; instances.

If your account has multiple plans and you want to filter by plan, set the `resource_plan_id` query parameter to one of the following values:

| Plan | `resource_plan_id` |
|------|---------------------|
| Premium | `7f666d17-7893-47d8-bf9d-2b2389fc4dfc` |
| Flex | `53bde9d3-cdbb-46f5-a98f-60ebcadf7260` |
| Pay-As-You-Go | `5304b575-3cff-4455-90dc-ae4367762093` |
| Open | `850b21a7-71de-4e53-9441-1abdd202f35d` |

> Each result includes `extensions` and `parameters` fields as described in [Get an instance](#get-an-instance).

Reviewer: Since `parameters` is explained in [Get an instance](#get-an-instance), can this be simplified by pointing users only to the field that matters to them?
Something like - Each result includes the same `extensions` fields as [Get an instance](#get-an-instance).

Pass a `parameters` JSON object in the request body with the fields you want to change, along with the header `"Content-Type: application/json"`. Omitted fields are left unchanged.

- `instance_limit_seconds` — Integer, or `null`. The usage time limit for the instance. See [Set instance allocation limits](/docs/guides/allocation-limits).
- `usage_allocation_seconds` — Integer, or `null`. The time allocated to this instance, used by the [fair-share scheduler](/docs/guides/fair-share-scheduler) to determine queue priority. See [Set instance allocation limits](/docs/guides/allocation-limits). Not applicable to Pay-As-You-Go instances.
- `backends` — Array of strings. The allowlist of backend names available to this instance. `["ANY"]` means all backends on the plan are available. `[]` means no backends are available.

<Admonition type="caution" title="Always include a unique timestamp">
  The API silently ignores the request if `parameters` is identical to the previous request. In the `parameters` object, always include a `timestamp` field set to the current time so each request is treated as unique.

> </Admonition>

Reviewer: Nice! and thank you! 💯

---
title: Use IBM Cloud Resource Controller API for instance management
description: Use the IBM Cloud Resource Controller REST API to get, list, update, and create Qiskit Runtime instances.

---

# Use IBM Cloud Resource Controller API for instance management

You can use the IBM Cloud&reg; Resource Controller REST API to programmatically get, create, and update [instances](/docs/guides/instances). 

All Resource Controller endpoints require that you authenticate by passing a header called `Authorization` with the bearer token. Refer to the [REST API setup guide](/docs/guides/cloud-setup-rest-api).

## Get an instance

Use the [`GET /v2/resource_instances/{crn}`](https://cloud.ibm.com/apidocs/resource-controller/resource-controller#get-resource-instance) endpoint to get information about a particular instance. The [CRN](/docs/guides/instances) must be URL-encoded in the path.

> In addition to the standard Resource Controller fields, the response includes Quantum-specific fields in `extensions`. Use `extensions` rather than `parameters` because `parameters` may include a subset of the full fields.

Reviewer 3: ```suggestion
In addition to the standard Resource Controller fields, the response includes quantum-specific fields in `extensions`. Use `extensions` rather than `parameters` because `parameters` might include a subset of the full fields.
```

The `extensions` object includes these fields:

- **`instance_limit_seconds`** — Integer, or `null`. The usage time limit for the instance. See [Set instance allocation limits](/docs/guides/allocation-limits).
- **`usage_allocation_seconds`** — Integer, or `null`. The time allocated to this instance, used by the [fair-share scheduler](/docs/guides/fair-share-scheduler) to determine queue priority. See [Set instance allocation limits](/docs/guides/allocation-limits).

> - **`backends`** — Array of strings. The allow-list of backend names available to this instance. `["ANY"]` means all backends on the plan are available (the default). `[]` means no backends are available.

Reviewer 3: ```suggestion
- **`backends`** — Array of strings. The allowlist of backend names available to this instance. `["ANY"]` means all backends on the plan are available (the default). `[]` means no backends are available.
```

## Get a list of all instances

> Use the [`GET /v2/resource_instances`](https://cloud.ibm.com/apidocs/resource-controller/resource-controller#list-resource-instances) endpoint to get a list of all your instances. Set the `resource_id` query parameter to `b6049020-80f4-11eb-a0f7-e35ec9b4054f` to filter to IBM Quantum instances.

Reviewer 3: ```suggestion
Use the [`GET /v2/resource_instances`](https://cloud.ibm.com/apidocs/resource-controller/resource-controller#list-resource-instances) endpoint to get a list of all your instances. Set the `resource_id` query parameter to `b6049020-80f4-11eb-a0f7-e35ec9b4054f` to filter to IBM Quantum&reg; instances.
```

| Pay-As-You-Go | `5304b575-3cff-4455-90dc-ae4367762093` |
| Open | `850b21a7-71de-4e53-9441-1abdd202f35d` |

Each result includes `extensions` and `parameters` fields as described in [Get an instance](#get-an-instance).

## Update an instance

Use the [`PATCH /v2/resource_instances/{crn}`](https://cloud.ibm.com/apidocs/resource-controller/resource-controller#update-resource-instance) endpoint to update the limit, allocation, and permitted backends for an instance. The [CRN](/docs/guides/instances) must be URL-encoded in the path.

Pass a `parameters` JSON object in the request body with the fields you want to change, along with the header `"Content-Type: application/json"`. Omitted fields are left unchanged.

- **`instance_limit_seconds`** — Integer, or `null`. The usage time limit for the instance. See [Set instance allocation limits](/docs/guides/allocation-limits).
- **`usage_allocation_seconds`** — Integer, or `null`. The time allocated to this instance, used by the [fair-share scheduler](/docs/guides/fair-share-scheduler) to determine queue priority. See [Set instance allocation limits](/docs/guides/allocation-limits). Not applicable to Pay-As-You-Go instances.

> - **`backends`** — Array of strings. The allow-list of backend names available to this instance. `["ANY"]` means all backends on the plan are available. `[]` means no backends are available.

Reviewer 3: ```suggestion
- **`backends`** — Array of strings. The allowlist of backend names available to this instance. `["ANY"]` means all backends on the plan are available. `[]` means no backends are available.
```

## Create a new instance

Use the [`POST /v2/resource_instances`](https://cloud.ibm.com/apidocs/resource-controller/resource-controller#create-resource-instance) endpoint to create (provision) a new instance. Pass a JSON body with the header `"Content-Type: application/json"`.

Required fields:

- **`name`** — A human-readable name for the instance.
- **`target`** — The region, such as `us-east` or `eu-de`.
- **`resource_plan_id`** — The plan for this instance. See the [plan ID table](#get-a-list-of-all-instances).
- **`resource_group`** — The [resource group](https://cloud.ibm.com/docs/account?topic=account-rgs) to use.

> You can also include a `parameters` object to set Quantum-specific values:

Reviewer 3: ```suggestion
You can also include a `parameters` object to set quantum-specific values:
```

- **`instance_limit_seconds`** — Integer, or `null`. The usage time limit for the instance. See [Set instance allocation limits](/docs/guides/allocation-limits).
- **`usage_allocation_seconds`** — Integer, or `null`. The time allocated to this instance, used by the [fair-share scheduler](/docs/guides/fair-share-scheduler) to determine queue priority. See [Set instance allocation limits](/docs/guides/allocation-limits). Not applicable to Pay-As-You-Go instances.

> - **`backends`** — Array of strings. The allow-list of backend names available to this instance. `["ANY"]` means all backends on the plan are available. `[]` means no backends are available.

Reviewer 3: ```suggestion
- **`backends`** — Array of strings. The allowlist of backend names available to this instance. `["ANY"]` means all backends on the plan are available. `[]` means no backends are available.
```
